README.md
ProxmoxVE API Client
====================

This **PHP 8+** library allows you to interact with your Proxmox server via API.

[![Latest Stable Version](https://poser.pugx.org/bennetgallein/proxmoxve/v/stable.svg)](https://packagist.org/packages/bennetgallein/proxmoxve)
[![Total Downloads](https://poser.pugx.org/bennetgallein/proxmoxve/downloads.svg)](https://packagist.org/packages/bennetgallein/proxmoxve)
[![Latest Unstable Version](https://poser.pugx.org/bennetgallein/proxmoxve/v/unstable.svg)](https://packagist.org/packages/bennetgallein/proxmoxve)
[![License](https://poser.pugx.org/bennetgallein/proxmoxve/license.svg)](https://packagist.org/packages/bennetgallein/proxmoxve)


Installation
------------

Recommended installation is using [Composer], if you do not have [Composer] what are you waiting?

In the root of your project execute the following:

```sh
$ composer require bennetgallein/proxmoxve ~v5.0
```

Or add this to your `composer.json` file:

```json
{
    "require": {
        "bennetgallein/proxmoxve": "~v5.0"
    }
}
```

Then perform the installation:
```sh
$ composer install --no-dev
```


Usage
-----

```php
<?php

// Require the autoloader
require_once 'vendor/autoload.php';

// Use the library namespace
use ProxmoxVE\Proxmox;

// Create your credentials array
$credentials = [
    'hostname' => 'proxmox.server.com',  // Also can be an IP
    'username' => 'root',
    'password' => 'secret',
];

// realm and port defaults to 'pam' and '8006' but you can specify them like so
$credentials = [
    'hostname' => 'proxmox.server.com',
    'username' => 'root',
    'password' => 'secret',
    'realm' => 'pve',
    'port' => '9009',
];

// you can also specify token-id and token-secret to use API tokens. They will be preffered over normal username/password
$credentials = [
    'hostname' => 'proxmox.server.com',
    'token-id' => 'root@pam!prod',
    'token-secret' => 'a806bdd1-de9a-4860-9ce0-bc427d3f2067'
]

// Then simply pass your credentials when creating the API client object.
$proxmox = new Proxmox($credentials);

$allNodes = $proxmox->get('/nodes');

print_r($allNodes);
```


Sample output:

```php
Array
(
    [data] => Array
        (
            [0] => Array
                (
                    [disk] => 2539465464
                    [cpu] => 0.031314446882002
                    [maxdisk] => 30805066770
                    [maxmem] => 175168446464
                    [node] => mynode1
                    [maxcpu] => 24
                    [level] =>
                    [uptime] => 139376
                    [id] => node/mynode1
                    [type] => node
                    [mem] => 20601992182
                )

        )

)
```

Using the MailGateway API
-------------------------

This library can be used to communicate with the [Proxmox MailGateway API](https://pmg.proxmox.com/pmg-docs/api-viewer/index.html) as well, by simply changing the `system` entry in the credentials array.

```php
<?php

$credentials = [
    'hostname' => 'proxmox.server.com',  // Also can be an IP
    'username' => 'root',
    'password' => 'secret',
    'system' => 'pmg', // Defaults to 'pve' for the PVE API. Use 'pmg' for the PMG API.
];

```

```
Using the backup Server API
-------------------------

This library can be used to communicate with the [Proxmox Backup Server API](https://pbs.proxmox.com/docs/api-viewer/index.html) as well, by simply changing the `system` entry in the credentials array.

```php
<?php

$credentials = [
    'hostname' => 'proxmox.server.com',  // Also can be an IP
    'username' => 'root',
    'password' => 'secret',
    'system' => 'pbs', // Defaults to 'pve' for the PVE API. Use 'pbs' for the PBS API.
];

```

Docs
----

On your proxmox client object you can use `get()`, `create()`, `set()` and `delete()` functions for all resources specified at [PVE2 API Documentation], params are passed as the second parameter in an associative array.

**What resources or paths can I interact with and how?**

In your proxmox server you can use the [pvesh CLI Tool](http://pve.proxmox.com/wiki/Proxmox_VE_API#Using_.27pvesh.27_to_access_the_API) to manage all the pve resources, you can use this library in the exact same way you would use the pvesh tool. For instance you could run `pvesh` then, as the screen message should say, you can type `help [path] [--verbose]` to see how you could use a path and what params you should pass to it. Be sure to [read about the pvesh CLI Tool](http://pve.proxmox.com/wiki/Proxmox_VE_API#Using_.27pvesh.27_to_access_the_API) at the [Proxmox wiki].

**How does the Proxmox API works?**

Consult the [ProxmoxVE API] article at the [Proxmox wiki].

**I feel you're hiding some cool tricks! I need more docs!**

See the project's [wiki](https://github.com/ZzAntares/ProxmoxVE/wiki) for a more
detailed documentation, it's all in there I swear it... I think.


License
-------

This project is released under the MIT License. See the bundled [LICENSE] file for details.


Want to contribute?
--------------------

Thank you! Take a look at the [CONTRIBUTING], you could easily set up a development environment to get you started in no time!


[LICENSE]:./LICENSE
[CONTRIBUTING]:./CONTRIBUTING.md
[PVE2 API Documentation]:http://pve.proxmox.com/pve-docs/api-viewer/index.html
[ProxmoxVE API]:http://pve.proxmox.com/wiki/Proxmox_VE_API
[Proxmox wiki]:http://pve.proxmox.com/wiki
[Composer]:https://getcomposer.org/


CHANGELOG.md
# [5.1.0](https://github.com/bennetgallein/ProxmoxVE/compare/v5.0.0...5.1.0) (2023-12-05)


### Features

* add parsing of port ([5b100a5](https://github.com/bennetgallein/ProxmoxVE/commit/5b100a5d7898978df61a05285ca259a0e7c6032c))

v3.0.0
------

- PHP 5.4 required.
- Proxmox API client logic rewritten (but the usage is pretty much the same).
- GuzzleHttp used for sending HTTP requests.
- Credentials default form is now an array not an object (read the docs).
- Better docs (compliant with this new release).


v2.1.1
------

- Small bug fixed, using an Eloquent model as a custom credentials object was not valid.
- Code coverage and code quality service integrated.


v2.1.0
------

- Now it is possible to specify the desired return type when creating Proxmox object, setters and getters for return type were added.
- Added docs about getters & setters and other tricks of the Proxmox client object.
- Now you can create a Proxmox API client object with a custom credentials object.


v2.0.0
------

- Namespace has changed, vendor name is not used anymore.
- The class that handles the API calls is now *Proxmox* not *ProxmoxVE*.
- Functions were renamed, instead of `get`, `post`, `put` and `delete` you use `get`, `create`, `set`, `delete` (to keep consistency with the *pvesh CLI Tool*).
- Documentation improved.
- Library now returns the errors messages instead of throwing an exception when a request was not successful.
- Source code refactored and respective tests were added.


v1.1.1
------

- Fixed bug when CURL was not enabled library used to crash.


v1.1.0
------

- Add check to see if CURL is enabled.
- Added capability to change Credentials after ProxmoxVE object creation.


v1.0.0
------

- Release of ProxmoxVE API Client.
- Use get, post, put, delete functions to request resources.
- Namespace includes vendor ZzAntares.


tests/TestCase.php
<?php
/**
 * This file is part of the ProxmoxVE PHP API wrapper library (unofficial).
 *
 * @copyright 2014 César Muñoz <zzantares@gmail.com>
 * @license http://opensource.org/licenses/MIT The MIT License.
 */

namespace ProxmoxVE;

use GuzzleHttp\Client;
use GuzzleHttp\Handler\MockHandler;
use GuzzleHttp\HandlerStack;
use GuzzleHttp\Psr7\Response;
use PHPUnit\Framework\TestCase as TTestCase;

/**
 * @author César Muñoz <zzantares@gmail.com>
 */
class TestCase extends TTestCase {
    protected function getMockProxmox($method = null, $return = null) {
        if ($method) {
            $proxmox = $this->getMockBuilder('ProxmoxVE\Proxmox')
                ->setMethods(array($method))
                ->disableOriginalConstructor()
                ->getMock();

            $proxmox->expects($this->any())
                ->method($method)
                ->will($this->returnValue($return));

        } else {
            $proxmox = $this->getMockBuilder('ProxmoxVE\Proxmox')
                ->disableOriginalConstructor()
                ->getMock();
        }

        return $proxmox;
    }


    protected function getProxmox($response) {
        $httpClient = $this->getMockHttpClient(true, $response);

        $credentials = [
            'hostname' => 'my.proxmox.tld',
            'username' => 'root',
            'password' => 'toor',
        ];

        return new Proxmox($credentials, null, $httpClient);
    }


    protected function getMockHttpClient($successfulLogin, $response = null) {
        if ($successfulLogin) {
            $stream = \GuzzleHttp\Psr7\Utils::streamFor('{"data":{"CSRFPreventionToken":"csrf","ticket":"ticket","username":"random"}}');
            $login  = new Response(202, ['Content-Length' => 0], $stream);
        } else {
            $login = new Response(400, ['Content-Length' => 0]);
        }

        $responseStream = \GuzzleHttp\Psr7\Utils::streamFor("{$response}");

        $mock = new MockHandler([
            $login,
            new Response(202, ['Content-Length' => 0], $responseStream),
        ]);


        $handler    = HandlerStack::create($mock);
        $httpClient = new Client(['handler' => $handler]);

        return $httpClient;
    }
}

tests/ProxmoxTest.php
<?php declare(strict_types=1);
/**
 * This file is part of the ProxmoxVE PHP API wrapper library (unofficial).
 *
 * @copyright 2014 César Muñoz <zzantares@gmail.com>
 * @license http://opensource.org/licenses/MIT The MIT License.
 */
namespace ProxmoxVE;

use GuzzleHttp\Exception\ClientException;
use GuzzleHttp\Exception\ConnectException;
use ProxmoxVE\Exception\MalformedCredentialsException;

/**
 * @author César Muñoz <zzantares@gmail.com>
 */
final class ProxmoxTest extends TestCase {

    public function testExceptionIsThrownIfBadParamsPassed() {
        $this->expectException(MalformedCredentialsException::class);
        new Proxmox('bad param');
    }


    public function testExceptionIsThrownWhenNonAssociativeArrayIsGivenAsCredentials() {
        $this->expectException(MalformedCredentialsException::class);
        new Proxmox([
            'root', 'So Bruce Wayne is alive? or did he died in the explosion?',
        ]);
    }


    /**
     * @expectedException \ProxmoxVE\Exception\MalformedCredentialsException
     */
    public function testExceptionIsThrownWhenIncompleteCredentialsArrayIsPassed() {
        $this->expectException(MalformedCredentialsException::class);
        new Proxmox([
            'username' => 'root',
            'password' => 'The NSA is watching us! D=',
        ]);
    }


    public function testGetCredentialsWithAllValues() {
        $ids = [
            'hostname' => 'some.proxmox.tld',
            'username' => 'root',
            'password' => 'I was here',
        ];

        $fakeAuthToken = new AuthToken('csrf', 'ticket', 'username');
        $proxmox       = $this->getMockProxmox('login', $fakeAuthToken);
        $proxmox->setCredentials($ids);

        $credentials = $proxmox->getCredentials();

        $this->assertEquals($credentials->getHostname(), $ids['hostname']);
        $this->assertEquals($credentials->getUsername(), $ids['username']);
        $this->assertEquals($credentials->getPassword(), $ids['password']);
        $this->assertEquals($credentials->getRealm(), 'pam');
        $this->assertEquals($credentials->getPort(), '8006');
        $this->assertEquals($credentials->getSystem(), 'pve');
    }


    public function testCredentialsWithMailGatewaySystem() {
        $ids = [
            'hostname' => 'some.proxmox.tld',
            'username' => 'root',
            'password' => 'I was here',
            'system'   => 'pmg',
        ];

        $fakeAuthToken = new AuthToken('csrf', 'ticket', 'username');
        $proxmox       = $this->getMockProxmox('login', $fakeAuthToken);
        $proxmox->setCredentials($ids);

        $credentials = $proxmox->getCredentials();

        $this->assertEquals($credentials->getSystem(), 'pmg');
    }

    public function testCredentialsWithBackupServerSystem() {
        $ids = [
            'hostname' => 'some.proxmox.tld',
            'username' => 'root',
            'password' => 'I was here',
            'system'   => 'pbs',
        ];

        $fakeAuthToken = new AuthToken('csrf', 'ticket', 'username');
        $proxmox       = $this->getMockProxmox('login', $fakeAuthToken);
        $proxmox->setCredentials($ids);

        $credentials = $proxmox->getCredentials();

        $this->assertEquals($credentials->getSystem(), 'pbs');
    }


    /**
     * @expectedException \Exception
     */
    public function testUnresolvedHostnameThrowsException() {
        $this->expectException(\Exception::class);
        $credentials = [
            'hostname' => 'proxmox.example.tld',
            'username' => 'user',
            'password' => 'pass',
        ];

        new Proxmox($credentials);
    }


    public function testLoginErrorThrowsException() {
        $this->expectException(ClientException::class);

        $credentials = [
            'hostname' => 'proxmox.server.tld',
            'username' => 'are not',
            'password' => 'valid folks!',
        ];

        $httpClient = $this->getMockHttpClient(false); // Simulate failed login

        new Proxmox($credentials, null, $httpClient);
    }

    public function testSetupWithPortInHost() {
        $credentials = [
            'hostname' => 'proxmox.server.tld:443',
            'username' => 'wu',
            'password' => 'tang'
        ];
        $creds       = new Credentials($credentials);
        $this->assertEquals($creds->getPort(), '443');
    }


    public function testGetAndSetResponseType() {
        $proxmox = $this->getProxmox(null);
        $this->assertEquals($proxmox->getResponseType(), 'array');

        $proxmox->setResponseType('json');
        $this->assertEquals($proxmox->getResponseType(), 'json');

        $proxmox->setResponseType('html');
        $this->assertEquals($proxmox->getResponseType(), 'html');

        $proxmox->setResponseType('extjs');
        $this->assertEquals($proxmox->getResponseType(), 'extjs');

        $proxmox->setResponseType('text');
        $this->assertEquals($proxmox->getResponseType(), 'text');

        $proxmox->setResponseType('png');
        $this->assertEquals($proxmox->getResponseType(), 'png');

        $proxmox->setResponseType('pngb64');
        $this->assertEquals($proxmox->getResponseType(), 'pngb64');

        $proxmox->setResponseType('object');
        $this->assertEquals($proxmox->getResponseType(), 'object');

        $proxmox->setResponseType('other');
        $this->assertEquals($proxmox->getResponseType(), 'array');
    }


    /**
     * @expectedException \InvalidArgumentException
     */
    public function testGetResourceWithBadParamsThrowsException() {
        $this->expectException(\InvalidArgumentException::class);
        $proxmox = $this->getProxmox(null);
        $proxmox->get('/someResource', 'wrong params here');
    }


    /**
     * @expectedException \InvalidArgumentException
     */
    public function testCreateResourceWithBadParamsThrowsException() {
        $this->expectException(\InvalidArgumentException::class);
        $proxmox = $this->getProxmox(null);
        $proxmox->create('/someResource', 'wrong params here');
    }


    /**
     * @expectedException \InvalidArgumentException
     */
    public function testSetResourceWithBadParamsThrowsException() {
        $this->expectException(\InvalidArgumentException::class);
        $proxmox = $this->getProxmox(null);
        $proxmox->set('/someResource', 'wrong params here');
    }


    /**
     * @expectedException \InvalidArgumentException
     */
    public function testDeleteResourceWithBadParamsThrowsException() {
        $this->expectException(\InvalidArgumentException::class);
        $proxmox = $this->getProxmox(null);
        $proxmox->delete('/someResource', 'wrong params here');
    }


    public function testGetResource() {
        $fakeResponse = <<<'EOD'
{"data":[{"disk":940244992,"cpu":0.000998615325210486,"maxdisk":5284429824,"maxmem":1038385152,"node":"office","maxcpu":1,"level":"","uptime":3296027,"id":"node/office","type":"node","mem":311635968}]}
EOD;
        $proxmox      = $this->getProxmox($fakeResponse);

        $this->assertEquals($proxmox->get('/nodes'), json_decode($fakeResponse, true));
    }
}

src/Proxmox.php
<?php

/**
 * This file is part of the ProxmoxVE PHP API wrapper library (unofficial).
 *
 * @copyright 2014 César Muñoz <zzantares@gmail.com>
 * @license http://opensource.org/licenses/MIT The MIT License.
 */

namespace ProxmoxVE;

use GuzzleHttp\Client;
use GuzzleHttp\Cookie\CookieJar;

use ProxmoxVE\Exception\AuthenticationException;
use ProxmoxVE\Exception\BadResponseException;

/**
 * ProxmoxVE class. In order to interact with the proxmox server, the desired
 * app's code needs to create and use an object of this class.
 *
 * @author César Muñoz <zzantares@gmail.com>
 */
class Proxmox {
    /**
     * @var \GuzzleHttp\Client()
     */
    private $httpClient;

    /**
     * Contains the proxmox server authentication data.
     *
     * @var \ProxmoxVE\Credentials
     */
    private $credentials;


    /**
     * Holds the response type used to requests the API, possible values are
     * json, extjs, html, text, png.
     *
     * @var string
     */
    private $responseType;


    /**
     * Holds the fake response type, it is useful when you want to get the JSON
     * raw string instead of a PHP array.
     *
     * @var string
     */
    private $fakeType;


    /**
     * Stores the ProxmoxVE user session such as ticket, username and csrf
     * prevention token, are all in there.
     *
     * @var \ProxmoxVE\AuthToken
     */
    private $authToken;


    /**
     * Constructor.
     *
     * @param mixed $credentials Credentials object or associative array holding
     *                           the login data.
     * @param string $responseType The response type that is going to be returned when doing requests.
     * @param \GuzzleHttp\Client $httpClient The HTTP client to be used to send requests over the network.
     *
     * @throws \ProxmoxVE\Exception\MalformedCredentialsException If bad args
     *                                                            supplied.
     * @throws \ProxmoxVE\Exception\AuthenticationException If given credentials
     *                                                      are not valid.
     */
    public function __construct(
        $credentials,
        $responseType = 'array',
        $httpClient = null
    ) {
        $this->setHttpClient($httpClient);

        // Set credentials and login to the Proxmox server.
        $this->setCredentials($credentials);

        $this->setResponseType($responseType);
    }


    /**
     * Send a request to a given Proxmox API resource.
     *
     * @param string $actionPath The resource tree path you want to request, see
     *                           more at http://pve.proxmox.com/pve2-api-doc/
     * @param array $params      An associative array filled with params.
     * @param string $method     HTTP method used in the request, by default
     *                           'GET' method will be used.
     *
     * @return \Psr\Http\Message\ResponseInterface
     *
     * @throws \InvalidArgumentException If the given HTTP method is not one of
     *                                   'GET', 'POST', 'PUT', 'DELETE',
     */
    private function requestResource($actionPath, $params = [], $method = 'GET') {
        $url = $this->getApiUrl() . $actionPath;

        if ($this->credentials->isUsingApiKeys()) {
            $headers = [
                'Authorization' => $this->credentials->getToken()
                // no need for CSRF token when using API Tokens
            ];
        } else {
            $cookies = CookieJar::fromArray([
                $this->getCookieName() => $this->authToken->getTicket(),
            ], $this->credentials->getHostname());

            $headers = [
                'CSRFPreventionToken' => $this->authToken->getCsrf(),
            ];
        }

        switch ($method) {
            case 'GET':
                return $this->httpClient->get($url, [
                    'verify'     => false,
                    'exceptions' => false,
                    'cookies'    => $cookies ?? [],
                    'query'      => $params,
                    'headers'    => $headers
                ]);
            case 'POST':
            case 'PUT':
            case 'DELETE':
                return $this->httpClient->request($method, $url, [
                    'verify'      => false,
                    'exceptions'  => false,
                    'cookies'     => $cookies ?? [],
                    'headers'     => $headers,
                    'form_params' => $params,
                ]);
            default:
                $errorMessage = "HTTP Request method {$method} not allowed.";
                throw new \InvalidArgumentException($errorMessage);
        }
    }


    /**
     * Parses the response to the desired return type.
     *
     * @param \Psr\Http\Message\ResponseInterface $response Response sent by the Proxmox server.
     *
     * @return mixed The parsed response, depending on the response type can be
     *               an array or a string.
     */
    private function processHttpResponse($response) {
        if ($response === null) {
            return null;
        }

        if ($response->getStatusCode() >= 400) {
            throw new BadResponseException($response->getReasonPhrase());
        }

        switch ($this->fakeType) {
            case 'pngb64':
                $base64 = base64_encode($response->getBody());
                return 'data:image/png;base64,' . $base64;
            case 'object': // 'object' not supported yet, we return array instead.
            case 'array':
                return json_decode($response->getBody(), true);
            default:
                return $response->getBody()->__toString();
        }
    }


    /**
     * Returns the name of the cookie that should be used to authenticate requests.
     *
     * @return string The required cookie for the system being used
     */
    private function getCookieName() {
        return match ($this->credentials->getSystem()) {
            'pmg' => "PMGAuthCookie",
            'pbs' => "PBSAuthCookie",
            default => 'PVEAuthCookie'
        };
    }


    /**
     * Sets the HTTP client to be used to send requests over the network, for
     * now Guzzle needs to be used.²
     *
     * @param \GuzzleHttp\Client $httpClient the client to be used
     */
    public function setHttpClient($httpClient = null) {
        $this->httpClient = $httpClient ?: new Client();
    }


    /**
     * Attempts to login using set credentials, if succeeded will return the
     * AuthToken used in all requests.
     *
     * @return \ProxmoxVE\AuthToken When successful login will return an
     *                              instance of the AuthToken class.
     *
     * @throws \ProxmoxVE\Exception\AuthenticationException If login fails.
     */
    public function login() {
        $loginUrl = $this->credentials->getApiUrl() . '/json/access/ticket';
        $response = $this->httpClient->post($loginUrl, [
            'verify'      => false,
            'exceptions'  => false,
            'form_params' => [
                'username' => $this->credentials->getUsername(),
                'password' => $this->credentials->getPassword(),
                'realm'    => $this->credentials->getRealm(),
            ],
        ]);

        $json = json_decode($response->getBody(), true);
        if (!$json['data']) {
            $error = 'Can not login using the provided credentials';
            throw new AuthenticationException($error);
        }

        return new AuthToken(
            $json['data']['CSRFPreventionToken'],
            $json['data']['ticket'],
            $json['data']['username']
        );
    }


    /**
     * Gets the Credentials object associated with this proxmox API instance.
     *
     * @return \ProxmoxVE\Credentials Object containing all proxmox data used to
     *                                connect to the server.
     */
    public function getCredentials() {
        return $this->credentials;
    }


    /**
     * Assign the passed Credentials object to the ProxmoxVE.
     *
     * @param \ProxmoxVE\Credentials|array $credentials A custom object holding credentials or a
     *                            Credentials object to assign.
     *
     * @throws \ProxmoxVE\Exception\AuthenticationException If can not login.
     */
    public function setCredentials($credentials) {
        if (!$credentials instanceof Credentials) {
            $credentials = new Credentials($credentials);
        }

        $this->credentials = $credentials;
        if (!$this->credentials->isUsingApiKeys()) {
            $this->authToken = $this->login();
        }
    }


    /**
     * Sets the response type that is going to be returned when doing requests.
     *
     * @param string $responseType One of json, html, extjs, text, png.
     */
    public function setResponseType($responseType = 'array') {
        $supportedFormats = array('json', 'html', 'extjs', 'text', 'png');

        if (in_array($responseType, $supportedFormats)) {
            $this->fakeType     = false;
            $this->responseType = $responseType;
        } else {
            switch ($responseType) {
                case 'pngb64':
                    $this->fakeType = 'pngb64';
                    $this->responseType = 'png';
                    break;
                case 'object':
                case 'array':
                    $this->responseType = 'json';
                    $this->fakeType = $responseType;
                    break;
                default:
                    $this->responseType = 'json';
                    $this->fakeType = 'array'; // Default format
            }
        }
    }


    /**
     * Returns the response type that is being used by the Proxmox API client.
     *
     * @return string Response type being used.
     */
    public function getResponseType() {
        return $this->fakeType ?: $this->responseType;
    }


    /**
     * GET a resource defined in the pvesh tool.
     *
     * @param string $actionPath The resource tree path you want to ask for, see
     *                           more at http://pve.proxmox.com/pve2-api-doc/
     * @param array $params      An associative array filled with params.
     *
     * @return array             A PHP array json_decode($response, true).
     *
     * @throws \InvalidArgumentException If given params are not an array.
     */
    public function get($actionPath, $params = []) {
        if (!is_array($params)) {
            $errorMessage = 'GET params should be an associative array.';
            throw new \InvalidArgumentException($errorMessage);
        }

        // Check if we have a prefixed '/' on the path, if not add one.
        if (substr($actionPath, 0, 1) != '/') {
            $actionPath = '/' . $actionPath;
        }

        $response = $this->requestResource($actionPath, $params);
        return $this->processHttpResponse($response);
    }


    /**
     * SET a resource defined in the pvesh tool.
     *
     * @param string $actionPath The resource tree path you want to ask for, see
     *                           more at http://pve.proxmox.com/pve2-api-doc/
     * @param array $params      An associative array filled with params.
     *
     * @return array             A PHP array json_decode($response, true).
     *
     * @throws \InvalidArgumentException If given params are not an array.
     */
    public function set($actionPath, $params = []) {
        if (!is_array($params)) {
            $errorMessage = 'PUT params should be an associative array.';
            throw new \InvalidArgumentException($errorMessage);
        }

        // Check if we have a prefixed '/' on the path, if not add one.
        if (substr($actionPath, 0, 1) != '/') {
            $actionPath = '/' . $actionPath;
        }

        $response = $this->requestResource($actionPath, $params, 'PUT');
        return $this->processHttpResponse($response);
    }


    /**
     * CREATE a resource as defined by the pvesh tool.
     *
     * @param string $actionPath The resource tree path you want to ask for, see
     *                           more at http://pve.proxmox.com/pve2-api-doc/
     * @param array $params      An associative array filled with POST params
     *
     * @return array             A PHP array json_decode($response, true).
     *
     * @throws \InvalidArgumentException If given params are not an array.
     */
    public function create($actionPath, $params = []) {
        if (!is_array($params)) {
            $errorMessage = 'POST params should be an asociative array.';
            throw new \InvalidArgumentException($errorMessage);
        }

        // Check if we have a prefixed '/' on the path, if not add one.
        if (substr($actionPath, 0, 1) != '/') {
            $actionPath = '/' . $actionPath;
        }

        $response = $this->requestResource($actionPath, $params, 'POST');
        return $this->processHttpResponse($response);
    }


    /**
     * DELETE a resource defined in the pvesh tool.
     *
     * @param string $actionPath The resource tree path you want to ask for, see
     *                           more at http://pve.proxmox.com/pve2-api-doc/
     * @param array $params      An associative array filled with params.
     *
     * @return array             A PHP array json_decode($response, true).
     *
     * @throws \InvalidArgumentException If given params are not an array.
     */
    public function delete($actionPath, $params = []) {
        if (!is_array($params)) {
            $errorMessage = 'DELETE params should be an associative array.';
            throw new \InvalidArgumentException($errorMessage);
        }

        // Check if we have a prefixed '/' on the path, if not add one.
        if (substr($actionPath, 0, 1) != '/') {
            $actionPath = '/' . $actionPath;
        }

        $response = $this->requestResource($actionPath, $params, 'DELETE');
        return $this->processHttpResponse($response);
    }


    // Later on below this line we'll move this logic to another place?


    /**
     * Returns the proxmox API URL where requests are sended.
     * Sample value: https://my-proxmox:8006/api2/json
     *
     * @return string Proxmox API URL.
     */
    public function getApiUrl() {
        return $this->credentials->getApiUrl() . '/' . $this->responseType;
    }


    /**
     * Retrieves the '/access' resource of the Proxmox API resources tree.
     *
     * @return mixed The processed response, can be an array, string or object.
     */
    public function getAccess() {
        return $this->get('/access');
    }


    /**
     * Retrieves the '/cluster' resource of the Proxmox API resources tree.
     *
     * @return mixed The processed response, can be an array, string or object.
     */
    public function getCluster() {
        return $this->get('/cluster');
    }


    /**
     * Retrieves the '/nodes' resource of the Proxmox API resources tree.
     *
     * @return mixed The processed response, can be an array, string or object.
     */
    public function getNodes() {
        return $this->get('/nodes');
    }


    /**
     * Retrieves the '/pools' resource of the Proxmox API resources tree.
     *
     * @return mixed The processed response, can be an array, string or object.
     */
    public function getPools() {
        return $this->get('/pools');
    }


    /**
     * Creates a pool resource inside the '/pools' resources tree.
     *
     * @param array $poolData An associative array filled with POST params
     *
     * @return mixed The processed response, can be an array, string or object.
     */
    public function createPool($poolData) {
        if (!is_array($poolData)) {
            throw new \InvalidArgumentException('Pool data needs to be array');
        }

        return $this->create('/pools', $poolData);
    }


    /**
     * Retrieves all the storage found in the Proxmox server, or only the ones
     * matching the storage type provided if any.
     *
     * @param string $type the storage type.
     *
     * @return mixed The processed response, can be an array, string or object.
     */
    public function getStorages($type = null) {
        if ($type === null) {
            return $this->get('/storage');
        }

        $supportedTypes = array(
            'lvm',
            'nfs',
            'dir',
            'zfs',
            'rbd',
            'iscsi',
            'sheepdog',
            'glusterfs',
            'iscsidirect',
        );

        if (in_array($type, $supportedTypes)) {
            return $this->get('/storage', array(
                'type' => $type,
            ));
        }

        /* If type not found returns null */
        return null;
    }


    /**
     * Creates a storage resource using the passed data.
     *
     * @param array $storageData An associative array filled with POST params
     *
     * @return mixed The processed response, can be an array, string or object.
     */
    public function createStorage($storageData) {
        if (!is_array($storageData)) {
            $errorMessage = 'Storage data needs to be array';
            throw new \InvalidArgumentException($errorMessage);
        }

        /* Should we check the required keys (storage, type) in the array? */

        return $this->create('/storage', $storageData);
    }


    /**
     * Retrieves the '/version' resource of the Proxmox API resources tree.
     *
     * @return mixed The processed response, can be an array, string or object.
     */
    public function getVersion() {
        return $this->get('/version');
    }
}